accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,887 @@
+    `Set` or `Map` instance originally created by a list, set or map literal. It
+    is a potentially constant element if the expression is potentially constant
+    expression.
+
+*   An `ifElement` is a constant element if its condition is a constant
+    expression evaluating to a Boolean value and either:
+
+    *   If the condition evaluates to `true`, then when the "then" element is a
+        constant expression and any "else" element is a potentially constant
+        element.
+
+    *   If the condition evaluates to `false`, then when the "then" element is a
+        potentially constant element and any "else" element is a constant
+        element.
+
+*   An `ifElement` is potentially constant if its condition, "then" element, and
+    "else" element (if any) are potentially constant expressions.
+
+*   A `forElement` is never a constant or potentially constant element. A `for`
+    element cannot be used in constant collection literals.
+
+Also:
+
+*   It is a compile-time error if any element in a constant set or key in a
+    constant map does not have a primitive operator `==`.
+
+*   When evaluating a const collection literal, if another const collection has
+    previously been evaluated that:
+
+    *   is of the same kind (map, set, or list),
+    *   has the same type arguments,
+    *   and contains the same (identical) elements or key/value entries in the
+        same order,
+
+    then the current literal evaluates to the value of that previous literal.
+    *In other words, constants are canonicalized.* Note that maps and sets are
+    only canonicalized if they contain the same keys or elements *in the same
+    order*.
+
+## Dynamic Semantics
+
+Spread, `if`, and `for` behave similarly across all collection types. To
+simplify the spec, there is a single procedure used for all kinds of
+collections. This recursive procedure builds up either a *result* sequence of
+values or key-value pair map entries. Then a final step handles those
+appropriately for the given collection type.
+
+### To evaluate a collection `element`:
+
+1.  If `element` is an expression element:
+
+    1.  Evaluate the element's expression and append it to *result*.
+
+1.  Else, `element` is a `mapEntry` `keyExpression: valueExpression`:
+
+    1.  Evaluate `keyExpression` to a value *key*.
+
+    1.  Evaluate `valueExpression` to a value *value*.
+
+    1.  Append an entry *key*: *value* to *result*.
+
+1.  Else, if `element` is a spread element:
+
+    1.  Evaluate the spread expression to a value `spread`.
+
+    1.  If `element` is not null-aware and `spread` is `null`, throw a dynamic
+        exception.
+
+    1.  Else, if `element` is null-aware and `spread` is `null`, do nothing.
+
+    1.  Else, if the collection is a map:
+
+        1.  Evaluate `spread.entries.iterator` to a value `iterator`.
+
+        1.  Loop:
+
+            1.  If `iterator.moveNext()` returns `false`, exit the loop.
```

Comment:
(If it doesn't return a `bool`, it's a run-time error).

Reply:
Changed it to use Boolean conversion like elsewhere in the proposal.